File: src/rb/matrix_calculator.rb
require_relative "matrix_result"

class CurrentHolder
    def initialize(row, column, side)
        @row = @startRow = row
        @column = @startColumn = column
        @side = side
    end

    def row
        @row
    end

    def row=row
        @row = row
    end
    
    def column
        @column
    end

    def column=column
        @column = column
    end

    def startRow
        @startRow
    end

    def startRow=startRow
        @startRow = startRow
    end

    def startColumn
        @startColumn
    end

    def startColumn=startColumn
        @startColumn = startColumn
    end

    def ActualRow
        return Actual(@row)
    end

    def ActualColumn
        return Actual(@column)
    end

    private
    def Actual(value)
        if value < 0
            return value + @side
        elsif value >= @side
            return value - @side
        end

        return value
    end
end

class MatrixCalculator
    def self.Calculate(side)

        if side % 2 != 1
            raise "Side needs to be an odd number."
        end

        matrix = CreateEmptyMatrix(side)
        current = CurrentHolder.new(-(side -1) / 2, (side / 2).floor, side)

        for count in 1..side * side
            matrix[current.ActualRow][current.ActualColumn] = count

            if count % side != 0
                current.row += 1
                current.column += 1
            else
                current.startRow += 1
                current.startColumn -= 1
                current.row = current.startRow
                current.column = current.startColumn
            end
            
        end
        
        return MatrixResult.new(matrix, matrix[0].inject(:+), side)
    end

    private
    def self.CreateEmptyMatrix(side)
        return Array.new(side) { Array.new(side) }
    end
end

File: src/lua/matrixCalculator.lua
local module = {}

local CurrentHolder = {
	row = 0,
	startRow = 0,
	column = 0,
	startColumn = 0,
	side = 0,
}

function CurrentHolder:new(row, column, side)
	local o = {}
	setmetatable(o, self)
	self.__index = self
	self.row = row
	self.startRow = row
	self.column = column
	self.startColumn = column
	self.side = side
	return o
end

function CurrentHolder:next(count)
	if count % self.side ~= 0 then
		self.row = self.row + 1
		self.column = self.column + 1
	else
		self.startRow = self.startRow + 1
		self.row = self.startRow
		self.startColumn = self.startColumn - 1
		self.column = self.startColumn
	end
end

function CurrentHolder:actual(value)
	if value < 0 then
		return value + self.side + 1
	end
	if value >= self.side then
		return value - self.side + 1
	end
	return value + 1
end

function CurrentHolder:actualRow()
	return self:actual(self.row)
end

function CurrentHolder:actualColumn()
	return self:actual(self.column)
end

local function createEmtpyMatrix(side)
	local matrix = {}
	for rowIndex = 1, side do
		local row = {}
		for columnIndex = 1, side do
			row[columnIndex] = 0
		end
		matrix[rowIndex] = row
	end
	return matrix
end

local function testResult(matrix, expectedSum, side)
	local sum = 0
	for rowIndex = 1, side do
		sum = 0
		for columnIndex = 1, side do
			sum = sum + matrix[rowIndex][columnIndex]
		end
		if sum ~= expectedSum then
			return false
		end
	end

	for columnIndex = 1, side do
		sum = 0
		for rowIndex = 1, side do
			sum = sum + matrix[rowIndex][columnIndex]
		end
		if sum ~= expectedSum then
			return false
		end
	end

	sum = 0
	for diagonalIndex = 1, side do
		sum = sum + matrix[diagonalIndex][diagonalIndex]
	end
	return sum == expectedSum
end

function module.calculate(side)
	if side % 2 ~= 1 then
		error("Side needs to be an odd number.")
	end

	local expectedSum = math.floor(side * (side * side + 1) / 2)
	local matrix = createEmtpyMatrix(side)
	local current = CurrentHolder:new(-math.floor((side - 1) / 2), math.floor(side / 2), side)
	for count = 1, side * side do
		local actualRow = current:actualRow()
		local actualColumn = current:actualColumn()
		matrix[actualRow][actualColumn] = count
		current:next(count)
	end
	return {
		matrix = matrix,
		sum = expectedSum,
		side = side,
		success = testResult(matrix, expectedSum, side),
	}
end

return module
